include/dlaf/sender/transform_mpi.h

```diff
@@ -9,39 +9,26 @@
 //
 #pragma once
 
-#include <pika/unwrap.hpp>
 #include <type_traits>
 
 #include "dlaf/common/pipeline.h"
+#include "dlaf/common/unwrap.h"
 #include "dlaf/communication/communicator.h"
 #include "dlaf/sender/transform.h"
 #include "dlaf/sender/when_all_lift.h"
 
 namespace dlaf::comm::internal {
 
-template <class T>
-struct IsPromiseGuardCommunicator : std::is_same<T, dlaf::common::PromiseGuard<Communicator>> {};
-
-/// This helper unwraps a PromiseGuard<Communicator> and returns the reference to the Communicator
-/// inside. Lifetime of the PromiseGuard has to be ensured by the caller.
-template <typename T>
-decltype(auto) unwrapPromiseGuardCommunicator(T& t) {
-  if constexpr (IsPromiseGuardCommunicator<std::decay_t<T>>::value) {
-    return t.ref();
-  }
-  else {
-    return static_cast<T&>(t);
-  }
+/// This helper "consumes" a PromiseGuard<Communicator> ensuring that after this call the one
+/// passed as argument gets destroyed. All other types left as they are by the
+/// second overload.
+inline void consumePromiseGuardCommunicator(dlaf::common::PromiseGuard<Communicator>& pcomm) {
+  [[maybe_unused]] auto pcomm_local = std::move(pcomm);
 }
 
-/// This helper "consumes" a PromiseGuard<Communicator> ensuring that after this call the one
-/// passed as argument gets destroyed.
+/// \overload consumePromiseGuardCommunicator
 template <typename T>
-void consumePromiseGuardCommunicator(T& t) {
-  if constexpr (IsPromiseGuardCommunicator<std::decay_t<T>>::value) {
-    [[maybe_unused]] auto t_local = std::move(t);
-  }
-}
+void consumePromiseGuardCommunicator(T&) {}
```

Comment:
Note, the only reason I separated this into two overloads is to avoid having to instantiate `IsPromiseGuardCommunicator`. I haven't measured the difference and I suspect the win is small, so if someone prefers the `if constexpr` version let me know and I'll happily revert this. This change has no impact on behaviour.